pkg/server/plugin/nodeattestor/gcp/iit.go

```diff
@@ -0,0 +1,197 @@
+package gcp
+
+import (
+	"context"
+	"crypto/x509"
+	"encoding/json"
+	"encoding/pem"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"path"
+	"sync"
+	"time"
+
+	"github.com/hashicorp/hcl"
+
+	jwt "github.com/dgrijalva/jwt-go"
+	cgcp "github.com/spiffe/spire/pkg/common/plugin/gcp"
+	spi "github.com/spiffe/spire/proto/common/plugin"
+	"github.com/spiffe/spire/proto/server/nodeattestor"
+)
+
+const (
+	pluginName    = "gcp_iit"
+	audience      = "spire-gcp-node-attestor"
+	googleCertURL = "https://www.googleapis.com/oauth2/v1/certs"
+)
+
+type tokenKeyRetriever interface {
+	retrieveKey(token *jwt.Token) (interface{}, error)
+}
+
+type googlePublicKeyRetriever struct {
+	certificates map[string]*x509.Certificate
+	expirey      int64
+	mtx          *sync.Mutex
+}
+
+func (r *googlePublicKeyRetriever) retrieveKey(token *jwt.Token) (interface{}, error) {
+	if token.Header["kid"] == nil {
+		return nil, fmt.Errorf("Missing kid in identityToken header. Cannot verify token")
+	}
+	kid := token.Header["kid"].(string)
+	if _, ok := token.Method.(*jwt.SigningMethodHMAC); !ok {
+		return nil, fmt.Errorf("Unexpected signing method: %v", token.Header["alg"])
+	}
+
+	if r.expirey == 0 || time.Now().Unix() > r.expirey {
+		r.mtx.Lock()
+		defer r.mtx.Unlock()
+		err := r.downloadCertificates()
+		if err != nil {
+			return nil, err
+		}
+	}
+	return r.certificates[kid].PublicKey, nil
+}
+
+func (r *googlePublicKeyRetriever) downloadCertificates() error {
+	resp, err := http.Get(googleCertURL)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+	bytes, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return err
+	}
+
+	var data map[string]string
+	err = json.Unmarshal(bytes, &data)
+
+	for k, v := range data {
+		block, _ := pem.Decode([]byte(v))
+		cert, err := x509.ParseCertificate(block.Bytes)
+		if err != nil {
+			return err
+		}
+		r.certificates[k] = cert
+	}
+	t, err := time.Parse("Mon, 2 Jan 2006 15:04:05 MST", resp.Header["Expires"][0])
+	r.expirey = t.Unix()
+	return nil
+}
+
+type IITAttestorConfig struct {
+	TrustDomain        string   `hcl:"trust_domain"`
+	ProjectIDWhitelist []string `hcl:"projectid_whitelist`
+}
+
+type IITAttestorPlugin struct {
+	trustDomain        string
+	projectIDWhitelist []string
+	tokenKeyRetriever  tokenKeyRetriever
+	mtx                *sync.Mutex
+}
+
+func (p *IITAttestorPlugin) spiffeID(gcpAccountID string, gcpInstanceID string) *url.URL {
+	spiffePath := path.Join("spire", "agent", pluginName, gcpAccountID, gcpInstanceID)
+	id := &url.URL{
+		Scheme: "spiffe",
+		Host:   p.trustDomain,
```

Comment:
Are we missing a read lock here? The agent plugin takes one when beginning the FetchAttestationData method, but I don't see a similar one here on Attest.